### ISSUE TITLE
Auth onCreate and onDelete triggers for v2 functions

### ISSUE THREAD

**Issue by laurentpayot** [firebase/firebase-functions#1383]
Please add Authentication onCreate and onDelete triggers for v2 functions in the SDK, this is really missing. :pray: 

### Related issues

https://github.com/firebase/firebase-functions/issues/1213
https://github.com/firebase/firebase-functions/issues/1381

### Version info

**firebase-functions:**
[v4.3.1](https://github.com/firebase/firebase-functions/releases/tag/v4.3.1)

**firebase-tools:**
11.29.0

**Comment by jhuleatt** [comment]
Hey @laurentpayot, thanks for requesting! This is currently a limitation of all 2nd gen functions (Node or Python). A workaround for now is to use 1st gen functions for Auth `onCreate` and `onDelete` triggers. Note that [1st gen and 2nd gen functions can coexist side-by-side in the same file](https://firebase.google.com/docs/functions/2nd-gen-upgrade#the_migration_process).

**Comment by laurentpayot** [comment]
Thanks for your answer @jhuleatt. I did not know a mix of 1st and 2nd gen functions was possible. I will use this "workaround" for now. I’m still eagerly awaiting v2 auth triggers :wink:

**Comment by lpellegr** [comment]
Any updates on this?

**Comment by jjgriff93** [comment]
Seems to be the only ones missing now, do we have an ETA on when these will make it into v2?

**Comment by jacobg** [comment]
+1

**Comment by dhutaryan** [comment]
+1

**Comment by emilioiz** [comment]
+1

**Comment by quirkles** [comment]
+1

**Comment by amacapas** [comment]
+1

**Comment by mvergarair** [comment]
+1

**Comment by Martinnord** [comment]
+1

**Comment by kafkas** [comment]
+1

**Comment by sebastianrueckerai** [comment]
When can we expect this?`

**Comment by sveinlo** [comment]
+1

**Comment by muraterzioglu** [comment]
+1

**Comment by solarpush** [comment]
> I believe we can use a HTTP v2 function for blocking "Before the user is created" https://firebase.google.com/docs/auth/extend-with-blocking-functions?gen=2nd. It is a workaround for onCreate callback.

Yes, I try it but this trigger not working in emulators ... (or i don't find how to) and i believe acces right is limited (for interact with firestore or other) 

I any one know a solution for use blocking functions in emulators

**Comment by vazome** [comment]
+1 for Python support

**Comment by newcaiosantos** [comment]
+1

**Comment by yousefh409** [comment]
+1

**Comment by Seanitzel** [comment]
Will this be implemented at some point?
The docs say the following - 
```
Note: Cloud Functions for Firebase (2nd gen) does not provide support for the events and triggers described in this guide. Because 1st gen and 2nd gen functions can coexist side-by-side in the same source file, you can still develop and deploy this functionality together with 2nd gen functions.
```
But they still urge us to upgrade all our functions to v2...

**Comment by jay8t6** [comment]
Yes, please add this. I updated my firebase functions to the latest, now I get firebase.runWith is not a function error. How can you say it co-exists when it cant after you upgrade to the latest version of firebase-functions?

+1

**Comment by haybreaker** [comment]
+1

**Comment by enchorb** [comment]
+1, classis half-baked major upgrades by Firebase

**Comment by rublev** [comment]
the amount of time i wasted on this.... lmao 🤡🔫

**Comment by sak6lab** [comment]
+1

**Comment by Smrtnyk** [comment]
+1

**Comment by 10eputzen** [comment]
+1

**Comment by matheusbaumgart** [comment]
They can't co-exist anymore, please update ASAP.

**Comment by zoubaierChraiet** [comment]
+1

**Comment by ondrej-111** [comment]
> Yes, please add this. I updated my firebase functions to the latest, now I get firebase.runWith is not a function error. How can you say it co-exists when it cant after you upgrade to the latest version of firebase-functions?
> 
> +1

Always check change-log when upgrading lib...

You can still use V1 functions but they defaulted to V2: https://github.com/firebase/firebase-functions/releases/tag/v6.0.0
so import now looks like: `import { runWith } from 'firebase-functions/v1'`

**Comment by vendoshi** [comment]
+1

**Comment by PepeR0nni** [comment]
+1

**Comment by elucidsoft** [comment]
What would the point be in upgrading if they defaulted to v2, and I still have to call v1, but now I am just calling it differently? No thanks, I'll just stay on my current version until they can make v2 not a complete disaster.

**Comment by zak245** [comment]
+1

**Comment by troya2** [comment]
+1

**Comment by atsjo** [comment]
now support for 1.gen functions is removed from latest node version, according to https://cloud.google.com/static/functions/docs/concepts/execution-environment#node.js

This actually worked up until about december 10, but 2 days later 1.gen functions failed to deploy with node 22 runtime...
So now I can't deploy 1. and 2. gen function together anymore, unless I downgrade node runtime to 20...
What is the plan for this going forward?
1. add support for 1.gen in nodejs 22
2. add 2. gen entrypoints for onCreate and onDelete?

I can se that other runtimes also deprecate 1.gen support, so I assume this is eventually going away, meaning it must be alternative 2 above...

But what is the status concerning this, this issue has been open for 1.5 years now...?
If I want to use nodejs 22 on my 2.gen functions, I now have to split the deployment and use nodejs 20 for onCreate and onDelete...

**Comment by evdama** [comment]
Any progress on this? My `package.json` looks like this

```json
{
  "dependencies": {
    "firebase-admin": "^13.0.1",
    "firebase-functions": "^6.1.2",
    "firestore-store": "^2.0.2",
    "mailtrap": "^3.4.0"
  },
  "devDependencies": {
    "@types/node": "^22.10.2",
    "@types/nodemailer": "^6.4.17",
    "tslib": "^2.8.1",
    "typescript": "^5.7.2"
  },
  "engines": {
    "node": "22"
  },
  "main": "lib/index.js",
  "name": "functions",
  "private": true,
  "scripts": {
    "build": "tsc",
    "deploy": "firebase deploy --only functions",
    "logs": "firebase functions:log",
    "serve": "firebase serve --only functions,hosting",
    "shell": "firebase functions:shell",
    "start": "pnpm run shell"
  }
}
```

firebase CLI is at this version:
```
sa@Mac: ~/0/edm/functions  |development ↑2 U:4 ?:1 ✗| firebase --version
13.29.1
sa@Mac: ~/0/edm/functions  |development ↑2 U:4 ?:1 ✗|
```

The functions looks like this
```ts
import { getAuth } from 'firebase-admin/auth'
import { MailtrapClient } from 'mailtrap'
import * as functions from 'firebase-functions'

const mailtrapAPIKey = functions.params.defineSecret('MAILTRAP_API_KEY')

export const setCustomClaimsOnEDMSignUp = functions
  .region('us-central1')
  .auth.user()
  .onCreate(async (user, context) => {
    try {
         .... here's a bunch of custom code ...
    } catch (error) {
      ;(error: any) => {
        functions.logger.error('🔴 error inside setCustomClaimsOnEDMSignUp: ', error)
      }
    }
  })
```

After upgrading to v6 I'm I can't deploy anymore:
```
sa@Mac: ~/0/edm/functions  |development ↑2 U:5 ?:1 ✗| firebase deploy --only functions:setCustomClaimsOnEDMSignUp
(node:31862) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)

=== Deploying to 'prod-version-D4738'...

i  deploying functions
i  functions: preparing codebase default for deployment
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  functions: required API cloudfunctions.googleapis.com is enabled
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
✔  functions: required API cloudbuild.googleapis.com is enabled
i  functions: Loading and analyzing source code for codebase default to determine what to deploy
Serving at port 8934

(node:31944) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)

TypeError: functions.region is not a function
    at Object.<anonymous> (/Users/sa/0/edm/functions/lib/firebase/auth.js:9:6)
    at Module._compile (node:internal/modules/cjs/loader:1565:14)
    at Object..js (node:internal/modules/cjs/loader:1708:10)
    at Module.load (node:internal/modules/cjs/loader:1318:32)
    at Function._load (node:internal/modules/cjs/loader:1128:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:219:24)
    at Module.require (node:internal/modules/cjs/loader:1340:12)
    at require (node:internal/modules/helpers:138:16)
    at Object.<anonymous> (/Users/sa/0/edm/functions/lib/index.js:8:14)


Error: Functions codebase could not be analyzed successfully. It may have a syntax or runtime error

Having trouble? Try firebase [command] --help
sa@Mac: ~/0/edm/functions  |development ↑2 U:5 ?:1 ✗|
```

**Comment by vazome** [comment]
**To all potential viewers:**

If you haven’t contacted Firebase to raise your concern yet, please take a moment to do so. It would be helpful if you could specify whether (and how severely) the lack of Firebase Auth `onCreate` and `onDelete` triggers impact your business processes or development: [Contact Support](https://firebase.google.com/support/troubleshooter/contact)

Also, please cast a vote at UserVoice: [Auth onCreate and onDelete triggers for v2 functions](https://firebase.uservoice.com/forums/948424-general/suggestions/46611049-auth-oncreate-and-ondelete-triggers-for-v2-functio)

**Comment by evdama** [comment]
@colerogers do you have any updates on this issue?

**Comment by htuli-tuw** [comment]
When will this get resolved ? Our whole codebase is on v2 just for this one trigger, we need to use v1.

**Comment by nmbgeek** [comment]
+1

**Comment by atsjo** [comment]
now I was able to deploy gen1 functions with nodejs22 again:-) Not mentioned in the docs, but at least it makes it easier to use the gen1 auth triggers in a gen2 codebase...

**Comment by w3rafu** [comment]
:(

**Comment by evdama** [comment]
There's one good news since yesterday... it seems v1 functions will now work on node runtime 22 to as well (v1 was node v20 before) i.e. when you have v1 and v2 functions the deployment becomes easy again. The docs still say v1 functions can only run up to node v20 but that's not true anymore.

@atsjo can you confirm?

**Comment by shaffy9633** [comment]
bruh yall have to fix this. this is embarrassing

**Comment by sconix** [comment]
This starts to be a theme, the parameter support is half done, the support for auth triggers is half done the support for custom service account is half done, pretty much everything is only half finished regarding cloud run functions and at the same time Google is moving forward deprecating things and asking everyone to upgrade, well would love to, but can only do it for half of the functions due to these unfinished features.  Would be nice to at least know when we can expect full v2 functions support. Yes one can run v1 and v2 at the same time, but who really wants to maintain two different code bases.

**Comment by maciejgoscinski-latrace** [comment]
Any official statement from Google, before they deprecate us all to nonexistence?

**Comment by rublev** [comment]
> Any official statement from Google, before they deprecate us all to nonexistence?

>me being deprecated like
![](https://media.tenor.com/JBCYX4zLffgAAAAM/peace-disappear.gif)

**Comment by mqln** [comment]
Hey guys, friendly reminder that everyone is sad at you.

**Comment by gbourne1** [comment]
+1

**Comment by readyiman** [comment]
Oh, it's apparently not yet available. That's why AI answer keep looping.

> This is currently a limitation of all 2nd gen functions (Node or Python). A workaround for now is to use 1st gen functions for Auth `onCreate` and `onDelete` triggers. Note that [1st gen and 2nd gen functions can coexist side-by-side in the same file](https://firebase.google.com/docs/functions/2nd-gen-upgrade#the_migration_process).

I wonder using v1 as workaround can still be used 2 years after.. 
Especially after everyone asked to do migration.

**Comment by solarpush** [comment]
For byPass limitation you need to use dedicated oncall or onrequest fn for create , update , delete a user with server auth package and use you logic at this time.

The front auth package can only be used for signin or signout.

You can too use the signinwithcustomtoken fn in front if you want signin from server.

**Comment by xrlife** [comment]
Pretty crazy that this is still an issue. I have not seen any replies to any of these posts for so long. Does anyone even work on this stuff? Trying to build something here and we can't really run this stuff anymore. Just stop, why force us to upgrade when you can't even force yourself to maintain your libraries. Only the posters here will get this info because no one at Google could give a crap about our projects and our success. It is garbage and if you worked for any of us we would fire you for incompetence and lack of communication at a minimum. Someone is updated the messages telling us we need to upgrade our node, why not upgrade your stuff at the same time. Act like you are one of the largest companies in the world. Maybe focus on something other than AdWords for a bit. Try it out. You committed to use node, and you know Nodes update cycle. Who dropped the ball here? It takes you longer to update your library than the cycle to deprecate your library. Seems like 101 stuff here. Maybe the interns should not be handing this project anymore. Seriously disappointing, since you are releasing great new features for firebase this same time you are doing nothing for basic triggers like OnCreateUser. Who would need such an edge case method, right?

**Comment by kafkas** [comment]
Hey folks, just wanted to check if I’m missing something here. I’ve been seeing a lot of complaints about the lack of auth onCreate/onDelete triggers in v2, but isn’t it totally fine to just mix v1 and v2 functions?

We’ve upgraded all our other triggers (Firestore, callables, etc.) to v2, and we’re still using v1 just for the auth triggers until they’re supported in v2. This setup works fine with both Node.js 20 and 22:

```ts
import { runWith } from 'firebase-functions/v1';
import { onDocumentCreated } from 'firebase-functions/v2/firestore';

export const authTriggerV1 = runWith({ /* config */ })
  .auth.user()
  .onCreate(async user => {
    // ...
  });

export const firestoreTriggerV2 = onDocumentCreated('users/{userId}', async event => {
  // ...
});
```

It’s been working smoothly for us. Once v2 supports auth triggers, we’ll migrate those too. Is there a reason this isn’t a viable solution for others?

**Comment by davie-robertson** [comment]
...and for what it is worth, the can no longer run side-by-side with v2 functions - https://github.com/firebase/firebase-functions/issues/1653

**Comment by kafkas** [comment]
@davie-robertson Did you try deploying yourself? It is working for me. The issue you linked is from December 2024. It is possible they added support recently and haven't updated the docs. 

<img width="869" alt="Image" src="https://github.com/user-attachments/assets/8da47ed6-0e34-4276-9674-1058047c9ee1" />

**Comment by davie-robertson** [comment]
hi @kafkas ,

Yes, I am deploying via `firebase deploy --only functions`, as as per that issue, it's not working in `europe-west1`

![Image](https://github.com/user-attachments/assets/319cc108-b6b5-4fc6-9d9a-10014b212473)

**Comment by mgrazianoc** [comment]
Layoffs are so good, am I right folks?

**Comment by RohovDmytro** [comment]
+googol

**Comment by emaysyuk** [comment]
+1

**Comment by emaysyuk** [comment]
Failed to update v1 functions with Node 22 in europe-central2

<img width="806" alt="Image" src="https://github.com/user-attachments/assets/526d9221-3134-4c7d-947a-5053d8c9b1ac" />

**Comment by Dranaxel** [comment]
Currently at the same state as @emaysyuk 

We deployed our prod setup on node 18 with the onUserCreate/Delete hook, we recently moved to node 22 and the cli is not able to create the new functions since. I'm pretty worried if the function gets deleted we would not been able to recreate the functions and thus not be able to onboard new clients...

This is a huge red flag for keeping relying on firebase suite in the future :/

**Comment by danielreis58** [comment]
Should I use Supabase? 🤔

**Comment by mhit30** [comment]
this worked!

**Comment by daveryderoxford** [comment]
Can I clarify if v2 onUserCreated, onUserDeleted are currently supported by Firebase on server and emulator.   Gemini is convinced they have been supported since functions v4.2.0 but I am less so, especially not I have found this issue.

**Comment by mhit30** [comment]
> Can I clarify if v2 onUserCreated, onUserDeleted are currently supported by Firebase on server and emulator.   Gemini is convinced they have been supported since functions v4.2.0 but I am less so, especially not I have found this issue. 

Hi, I ran into this problem; I used v1 and it seemed to work. I used the firebase server though, not the emulator.

**Comment by omgovich** [comment]
Any plans on having onDelete and onCreate auth triggers as v2?
It is a huge blocker — v1 doesn't support Node22. At the same time, v2 doesn't have auth triggers. Some kind of a joke.

**Comment by RohovDmytro** [comment]
+1

**Comment by milos538** [comment]
+1
Please fix. Can not deploy old auth triggers on Node 22.

**Comment by vojdan** [comment]
> +1
> Please fix. Can not deploy old auth triggers on Node 22.

Same here.

**Comment by maremoto** [comment]
+1 
I want to migrate to v2 cloud functions with python before node 18 is deprecated but still blocked with this issue

**Comment by taeold** [comment]
Hi folks. We are definitely aware of this gap in v2 and actively working to fill it. Stay tuned.

**Comment by shaunbyrnes** [comment]
This is absurd how long those triggers have not been added - we have been waiting for this "update" for like 2 years since v2 was released lol. 

When is it being rolled out?

**Comment by OliverJEvans** [comment]
+1

**Comment by HM-Cloudamai** [comment]
Can we please have this already. wow.

**Comment by aleks-afanasyev-sotec** [comment]
+1

**Comment by Mitsu13Ion** [comment]
+1

**Comment by gogupaul123** [comment]
This is absolutely ridiculous. We NEED those in v2

**Comment by plackowski** [comment]
+1
Please fix. Can not deploy old auth triggers on Node 22.

**Comment by JoergTiedemann** [comment]
This issue is 2 year old and still open.
Can anyone of the contributors please leave comment about this issue and increase the priority please ?

**Comment by solarpush** [comment]
Or finally says you remove this feature and remove the documentation about.

> This issue is 2 year old and still open.
> Can anyone of the contributors please leave comment about this issue and increase the priority please ?

**Comment by shaunbyrnes** [comment]
This needs to be fixed........ @jhuleatt @google-oss-bot

**Comment by AdamKundracik** [comment]
+1

**Comment by AndRud** [comment]
+1 
Please fix

**Comment by npmcoding** [comment]
+1
This is the only way I know of to automatically create a document in a collect on user registration

**Comment by awais-100** [comment]
+1
It is difficult to use v1 functions inside v2 functions codebase with nodejs 22, because v1 functions don't created with nodejs 22.

**Comment by gogupaul123** [comment]
Would also be EXTREMELY helpful if, in a million years when auth functions get added to v2, we could get an onUpdated trigger as well, to track changes to the auth user

**Comment by careduz** [comment]
+1 for Auth onCreate/onDelete triggers in Gen2.

We have comprehensive user cleanup logic (team membership, auto-promotion, token deletion) that relies on the Auth onDelete trigger. With Node 18 EOL on Oct 31, 2025, we'll be unable to deploy updates to these critical functions.

The current workaround (client-side HTTP endpoints) introduces coordination complexity and loses the automatic trigger guarantee that makes  these functions reliable.

Would appreciate any timeline or updates on Gen2 Auth trigger support. Thanks for your work on Cloud Functions!

**Comment by eleventhaus** [comment]
+1 🙏🏼

**Comment by jadegopher** [comment]
Node 18 EOL is close (Oct 31 2025) and there is no workaround to handle user.OnCreate event in cloud functions V2. It's crucial to have this event in 2nd gen functions ASAP🙏🏼

 
> ⚠  functions: Runtime Node.js 18 was deprecated on 2025-04-30 and will be decommissioned on 2025-10-30, after which you will not be able to deploy without upgrading. Consider upgrading now to avoid disruption. See  https://cloud.google.com/functions/docs/runtime-support for full details on the lifecycle policy

**Comment by solarpush** [comment]
> Node 18 EOL is close (Oct 31 2025) and there is no workaround to handle user.OnCreate event in cloud functions V2. It's crucial to have this event in 2nd gen functions ASAP🙏🏼
> 
> > ⚠  functions: Runtime Node.js 18 was deprecated on 2025-04-30 and will be decommissioned on 2025-10-30, after which you will not be able to deploy without upgrading. Consider upgrading now to avoid disruption. See  https://cloud.google.com/functions/docs/runtime-support for full details on the lifecycle policy

don't hope for anything

**Comment by ricsantos** [comment]
v1 triggers work with node 20

**Comment by sami-kuiskumetsa** [comment]
Is there a technical limitation that blocks implementing this or is this just a very low priority task? It worries me that every time I create a new Firebase app there's less and less locations available that do serve the Gen1 functions. It's often a pain to try to choose the locations for all the products so that there wouldn't be unnecessary hop outside of the datacenter, and the limited selection of datacenters supporting gen1 is part of the problem. It seems likely that gen1 support will be phased out before we get these triggers into gen2.

I guess this is just the consequence of forcibly trying to combine all the similar offerings under the same umbrella. There is a danger in that though. Many nifty features that once drove developer adoption have been dropping over the years. Wouldn't it make sense in both technical and business perspective to have these products be as easy to use together as possible?

**Comment by kdawgwilk** [comment]
Whats crazy is that this is the most upvoted issue by more than double the next highest upvoted issue and we are not seeing any communication from the maintainers 🤯 @taeold Any updates on this?

**Comment by solarpush** [comment]
And nobody thinks to contribute directly ? 😅


> Whats crazy is that this is the most upvoted issue by more than double the next highest upvoted issue and we are not seeing any communication from the maintainers 🤯 @taeold Any updates on this?

**Comment by kdawgwilk** [comment]
Alright I'll bite, here is a PR for implementing both of these missing functions in v2 #1762

**Comment by kdawgwilk** [comment]
Well I tried y'all https://github.com/firebase/firebase-functions/pull/1762#issuecomment-3504448241

**Comment by taeold** [comment]
Hey folks. Our team is very aware that lack of Firebase Auth trigger support in v2 function is frustrating.

The feature requires work across several teams across Google Cloud Platform, and I hope we can share more news about the feature soon. Stay tuned.

**Comment by kdawgwilk** [comment]
NodeJS 24 function runtime is now GA but only supports running v2 functions, at least thats what the docs say

https://docs.cloud.google.com/functions/docs/runtime-support#node.js
<img width="882" height="241" alt="Image" src="https://github.com/user-attachments/assets/cf9e75ca-3922-48f3-becc-15f70bf7ef65" />

**Comment by shep920** [comment]
As the other user above me said, g1 versions cant be deployed as easily. Is there any updates on v2 yet?

**Comment by jsteffensen** [comment]
Come on man! 
Are we going to get to 3 yrs and no fix!? Embarrassing...

**Comment by laurentpellegrino** [comment]
We're going to move to [Zitadel](https://github.com/zitadel/zitadel) as soon as possible due to this joke.

**Comment by gbourne1** [comment]
Firebase team can we please get an update - 3 years and counting on this issue.

**Comment by cabljac** [comment]
Hi all, 

Aware that there is significant demand for this feature. I will escalate this to the team and will provide an update as soon as it's available

**Comment by DavidPluxia** [comment]
omg finally please. How can the main feature people use the cloud functions for not be Gen2

**Comment by cmjordan42** [comment]
Is Firebase Auth being deprecated? The original release of Gen2 indicated that "auth is still being worked on" but 2.5 years later, not a peep, so that couldn't have been the truth. Now it's clear that even the work to support Auth on modern LTS versions of Node is not being done.

Node 20 is being retired (and Firebase now warns on this) yet Auth Gen1 doesn't support Node 24. So the only version of Node that we can use is Node 22 if you are a user of Auth. Node 22 is EOL in about a year.

Google has a bad reputation for suddenly stopping support for their products... just tell people the truth so they can get a head start on the work to move away from Firebase.

**Comment by jacobg** [comment]
@taeold Please provide status update on Firebase Auth triggers on GCF gen2.

**Comment by timmolter** [comment]
I had Claude analyze this to see if there was some better workaround for now:

Here's the full picture after researching the issue and PR #1762:

**There is no better workaround.** The problem is deeper than the Firebase Functions SDK — the Firebase maintainer taeold explicitly closed PR #1762 stating:

"this PR won't work because **Google Cloud Functions does not support 2nd Gen functions triggered from Firebase Auth/GCIP yet**"

The SDK can't implement it because the GCF infrastructure layer itself doesn't support GCIP-triggered gen2 functions. It requires cross-team work at Google Cloud Platform, not just a Firebase SDK change.

v1+v2 mix (your current approach) — the only officially supported workaround. Firebase themselves recommend it.

**The looming risk** is Node 22 going EOL in ~1 year, and Node 24 only runs gen2. GCF gen1 may not get Node 24 support, which means you'd be stuck on Node 22 for all functions unless Firebase ships the gen2 auth triggers before then. Since the issue is 3 years old with no ETA, this is a real concern.

**Bottom line**: Your v1+v2 mix is correct and pragmatic. It's not an ugly hack — it's the officially recommended pattern. There's nothing meaningfully better available right now.

**Comment by cabljac** [comment]
Hi all,

From what I understand, auth trigger support is on the roadmap for v2. The team are waiting on an upstream blocker before they can roll anything out. I'll continue to provide updates as I get them.

Thanks for your patience here.

**Comment by cmjordan42** [comment]
@cabljac thanks for getting back to us with an update. Since you didn't link it, I presume that blocker is not being tracked publicly on Github. Is there any additional info to identify what that upstream blocker is or a contact for GCP customers to voice their support to ensure it gets completed?

**Comment by cabljac** [comment]
@cmjordan42  Thanks, that's correct - not tracked on here and an internal infra thing I believe. I have a call with the team today and will be discussing this issue. I will ask about another contact point for you.

**Comment by cmjordan42** [comment]
What does the phase:raised tag that was added mean?

We're a month away from being unable to deploy. If I don't see any concrete movement on this in the next week, I'm going to have to migrate away.

**Comment by nilsreichardt** [comment]
> We're a month away from being unable to deploy. If I don't see any concrete movement on this in the next week, I'm going to have to migrate away.

The deprecation phase starts in a month. **You will be able to deploy Cloud Functions.** From 30 October 2026 onwards, you will no longer be able to deploy Cloud Functions with Node 20.

Source: https://docs.cloud.google.com/run/docs/runtime-support#deprecation_and_decommissioning

Also 1st gen Run functions still support Node 22. Deployment for Node 22 is supported until 30 Octoboer 2027. See https://docs.cloud.google.com/functions/docs/runtime-support#node.js.

**Comment by thomas-Dooda** [comment]
PoC in europe-west1,
  v1 trigger is our only non-EU component, blocks full EU data residency for GDPR.

**Comment by th0rgall** [comment]
@thomas-Dooda I have my v1 auth onCreate and onDelete hosted in europe-west1, that should already be possible?

**Comment by thomas-Dooda** [comment]
@th0rgall Thanks for the pointer! I tried looking into this but it seems risky in production. While `.region("europe-west1").auth.user().onCreate()` deploys fine, Firebase Auth events are tied to `us-central1` internally. There is a related report in #1409 where a `beforeCreate` blocking function deployed to `europe-west1` executes but silently fails to create the user in production (works in emulators though). So the region syntax is accepted but the underlying event routing may not actually work reliably outside `us-central1`.

**Comment by OutdatedGuy** [comment]
I think you all are referring to [this documentation](https://docs.cloud.google.com/functions/docs/runtime-support#node.js) which states that Node.js 24 runtime doesn't support 1st Gen functions.

<img width="927" height="340" alt="Image" src="https://github.com/user-attachments/assets/54c2b83c-8736-4c5a-9b7d-5413004aa0a1" />

<br>

But I think the documentation is wrong or missing something as I've been using 1st Gen Auth `onCreate` function on Node.js 24 for the past 6 months. Just checked it in my console: https://console.cloud.google.com/functions/list

<img width="337" height="68" alt="Image" src="https://github.com/user-attachments/assets/617f39ff-f6a4-4b4f-ab3d-aa0a7a0f7a1e" />

**Comment by th0rgall** [comment]
@OutdatedGuy interesting, for me it says `Node.js 22` for `Firebase Authentication (Preview)` on that page, and when I tried to deploy them with Node v24 using `firebase-tools` in January, I got an error that blocked the deploy, see this issue https://github.com/firebase/firebase-functions/issues/1805. I still get this error when trying to deploy today using the latest `firebase-tools@15.15.0`.

My functions are deployed in `europe-west1` though, and as @thomas-Dooda suggests above that may not be supported as well as `us-central1`. Where are your v1 triggers deployed? Or perhaps there is a difference between creating and updating a v1 auth trigger function.

In any case, this issue is about support for auth triggers in v2 functions. Node v24 support for v1 functions is maybe better discused in https://github.com/firebase/firebase-functions/issues/1805, even though the two issues are definitely related.

**Comment by OutdatedGuy** [comment]
I'm deploying the v1 auth in `asia-south1` region.

**Comment by Mistic92** [comment]
Any update on that?

**Comment by shaunbyrnes** [comment]
@cabljac When is the v2 trigger being released? It has been like 3 years now and still nothing...

**Comment by cabljac** [comment]
Hi there! The only information that I can provide is that this is being actively worked on at the moment, but I can't provide any concrete timeline yet. Thanks for you patience

**Comment by HSCOGT** [comment]
+1

**Comment by RohovDmytro** [comment]
Really?..

+1

**Comment by malcolm98** [comment]
+1

**Comment by NNskelly** [comment]
+1. Last few times I've tried to deploy our last remaining v1 functions - all auth triggers - on node22+ I got errors that node22 did not support v1 functions, to the point where I've now begun to note on our project config comments that we can't accept the current node defaults on new instances and need to manually specify node20 target.

**Comment by emaysyuk** [comment]
+1

**Comment by shaunbyrnes** [comment]
Whats up with this @firebase-ops ...?

**Comment by DexEze** [comment]
+1

**Comment by JoergTiedemann** [comment]
> Hi there! The only information that I can provide is that this is being actively worked on at the moment, but I can't provide any concrete timeline yet. Thanks for you patience

Is there anything new about this issue ? Is there a roughly time schedule this will be fixed after now 3 years ?

**Comment by okydk** [comment]
Is today the day? 🤯
https://firebase.google.com/support/releases#cloudfunctions-2nd-gen-auth-triggers

**Comment by JoergTiedemann** [comment]
yes it seems so, unbelievable that an issue from April 2023 seems to be solved
Firebase functions 7.4.0 should have on usercreateTrigger as v2 function
Did you have test it ?

**Comment by Hypenate** [comment]
I tried it, but the emulator is not ready it seems.

Update:
It has been fixed and it's ready in the latest release:
https://github.com/firebase/firebase-tools/releases/tag/v15.30.2

**Comment by DexEze** [comment]
I transitioned my v1 oncreate over to the v2 wrapper and works for me :)

**Comment by cmjordan42** [comment]
Did you not test it? Or did you find a way to get the emulator to work with it?

**Comment by IzaakGough** [comment]
Closing as v2 auth triggers are now available in 7.4.0.

**Comment by IzaakGough** [comment]
@Hypenate Thank you for raising the emulator issue, we will review as soon as possible.

**Comment by cmjordan42** [comment]
This is a pretty unacceptable definition of "available"... it has a major blocking defect that were surprisingly never caught in testing. Is there a Google Firebase product manager on this thread with a higher definition of complete?

**Comment by DexEze** [comment]
Ahh, I used it in my remote dev environment, not in the emulator. I see also the emulator doesn't work with it.

**Comment by cabljac** [comment]
Hi @cmjordan42  not a Google PM here but I help maintain things, i've reopened this for now and we'll raise the emulator issue with the team.

**Comment by cmjordan42** [comment]
Many thanks!

**Comment by mgrazianoc** [comment]
We won folks!

**Comment by cabljac** [comment]
I'm going to tentatively close this issue now 😄  Lots of thanks to you all for your patience with this!

If you encounter problems with the v2 auth triggers feel free to open a fresh issue, i'll keep an eye out for them so we can prioritise accordingly

**Comment by shaunbyrnes** [comment]
@cabljac how do we get access to use the new functions while in preview? 

We get Module '"firebase-functions/v2/identity"' has no exported member 'onUserCreated'.ts(2305) when trying to import the new function per docs here https://firebase.google.com/docs/functions/auth-events

Should be latest versions (in package.json):

    "firebase-admin": "^14.4.0",
    "firebase-functions": "^7.4.0",
